src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -5133,6 +5133,23 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
 EXTERN_C void JIT_PInvokeBegin(InlinedCallFrame* pFrame);
 EXTERN_C void JIT_PInvokeEnd(InlinedCallFrame* pFrame);
 
+EXTERN_C void JIT_ReversePInvokeEnter()
+{
+    Thread* thread = GetThreadNULLOk();
+    if (thread == NULL)
+        thread = CreateThreadBlockThrow();
+
+    thread->DisablePreemptiveGC();
+}
+
+EXTERN_C void JIT_ReversePInvokeExit()
+{
+    Thread* thread = GetThread();
```

Comment:
It may be a good idea to cache the `Thread*` in the PInvoke frame allocated by the JIT. Fetching it from stack should be faster than fetching it from thread local.

Reply:
Sounds good. Does this mean I should bring back the `ReversePInvokeFrame` I proposed?

Reply:
You just need to store a single `Thread*` pointer on the stack for this. The `ReversePInvokeFrame` you have proposed was much bigger than that.

You can wrap it in a struct if you would like for clarity. But it should be just a single `Thread*` pointer.